File: backend/src/test/java/edu/cit/cabasag/barberconnect/feature/auth/AuthServiceTest.java
package edu.cit.cabasag.barberconnect.feature.auth;

import edu.cit.cabasag.barberconnect.dto.request.LoginRequest;
import edu.cit.cabasag.barberconnect.dto.response.AuthResponse;
import edu.cit.cabasag.barberconnect.factory.UserFactory;
import edu.cit.cabasag.barberconnect.security.JwtUtil;
import edu.cit.cabasag.barberconnect.service.AuthService;
import edu.cit.cabasag.barberconnect.service.UserService;
import com.google.firebase.auth.FirebaseAuth;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.util.ReflectionTestUtils;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

/**
 * Unit Tests — Auth Feature Slice
 * TC-AUTH-01 through TC-AUTH-04
 *
 * Password verification uses the Firebase Auth REST API via RestTemplate.
 * We inject a mock RestTemplate via ReflectionTestUtils so no real HTTP call
 * is made during tests.
 */
@ExtendWith(MockitoExtension.class)
@DisplayName("Auth Feature — AuthService Tests")
class AuthServiceTest {

    @Mock private UserService  userService;
    @Mock private FirebaseAuth firebaseAuth;
    @Mock private JwtUtil      jwtUtil;
    @Mock private UserFactory  userFactory;
    @Mock private RestTemplate restTemplate;

    @InjectMocks private AuthService authService;

    private User mockUser;

    @BeforeEach
    void setUp() {
        mockUser = new User();
        mockUser.setUser_id("uid-001");
        mockUser.setFirstName("Juan");
        mockUser.setLastName("dela Cruz");
        mockUser.setEmail("juan@test.com");
        mockUser.setRole(User.UserRole.CUSTOMER);
        mockUser.setIsActive(true);

        // Inject the dummy API key and the mock RestTemplate into the service
        ReflectionTestUtils.setField(authService, "firebaseWebApiKey", "test-api-key");
        ReflectionTestUtils.setField(authService, "restTemplate", restTemplate);
    }

    // ── TC-AUTH-01 ────────────────────────────────────────────────────────────

    @Test
    @DisplayName("TC-AUTH-01: login() returns AuthResponse when Firebase verifies credentials")
    void login_validCredentials_returnsAuthResponse() {
        // Arrange — RestTemplate returns a 200 OK (password accepted)
        when(restTemplate.postForEntity(anyString(), any(), eq(Map.class)))
                .thenReturn(ResponseEntity.ok(Map.of("idToken", "firebase-id-token")));
        when(userService.findByEmail("juan@test.com")).thenReturn(Optional.of(mockUser));
        when(jwtUtil.generateToken(anyString(), anyString(), anyString()))
                .thenReturn("mocked.jwt.token");

        LoginRequest request = new LoginRequest();
        request.setEmail("juan@test.com");
        request.setPassword("password123");

        // Act
        AuthResponse result = authService.login(request);

        // Assert
        assertThat(result).isNotNull();
        assertThat(result.getEmail()).isEqualTo("juan@test.com");
        assertThat(result.getToken()).isEqualTo("mocked.jwt.token");
        verify(jwtUtil).generateToken(eq("uid-001"), eq("juan@test.com"), eq("CUSTOMER"));
    }

    // ── TC-AUTH-02 ────────────────────────────────────────────────────────────

    @Test
    @DisplayName("TC-AUTH-02: login() throws RuntimeException when user not found in Firestore")
    void login_userNotFound_throwsException() {
        // Arrange — Firebase accepts the password but user isn't in Firestore
        when(restTemplate.postForEntity(anyString(), any(), eq(Map.class)))
                .thenReturn(ResponseEntity.ok(Map.of("idToken", "firebase-id-token")));
        when(userService.findByEmail(anyString())).thenReturn(Optional.empty());

        LoginRequest request = new LoginRequest();
        request.setEmail("notfound@test.com");
        request.setPassword("password");

        // Act & Assert
        assertThatThrownBy(() -> authService.login(request))
                .isInstanceOf(RuntimeException.class)
                .hasMessageContaining("Invalid email or password");
    }

    // ── TC-AUTH-03 ────────────────────────────────────────────────────────────

    @Test
    @DisplayName("TC-AUTH-03: login() throws RuntimeException when account is deactivated")
    void login_deactivatedAccount_throwsException() {
        // Arrange
        mockUser.setIsActive(false);
        when(restTemplate.postForEntity(anyString(), any(), eq(Map.class)))
                .thenReturn(ResponseEntity.ok(Map.of("idToken", "firebase-id-token")));
        when(userService.findByEmail("juan@test.com")).thenReturn(Optional.of(mockUser));

        LoginRequest request = new LoginRequest();
        request.setEmail("juan@test.com");
        request.setPassword("password123");

        // Act & Assert
        assertThatThrownBy(() -> authService.login(request))
                .isInstanceOf(RuntimeException.class)
                .hasMessageContaining("deactivated");
    }

    // ── TC-AUTH-04 ────────────────────────────────────────────────────────────

    @Test
    @DisplayName("TC-AUTH-04: login() throws RuntimeException when Firebase rejects the password")
    void login_wrongPassword_throwsException() {
        // Arrange — Firebase returns 400 Bad Request for wrong password
        when(restTemplate.postForEntity(anyString(), any(), eq(Map.class)))
                .thenThrow(new HttpClientErrorException(HttpStatus.BAD_REQUEST,
                        "INVALID_PASSWORD"));

        LoginRequest request = new LoginRequest();
        request.setEmail("juan@test.com");
        request.setPassword("wrongpassword");

        // Act & Assert
        assertThatThrownBy(() -> authService.login(request))
                .isInstanceOf(RuntimeException.class)
                .hasMessageContaining("Invalid email or password");

        // Firestore should never be queried if Firebase rejects the password
        verify(userService, never()).findByEmail(anyString());
    }
}

File: web/barberconnect-frontend/src/test/authService.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { authService } from '../services/authService';

// ── Mock the api module ───────────────────────────────────────────────────────
vi.mock('../services/api', () => {
  const mockApi = {
    post: vi.fn(),
    get: vi.fn(),
    interceptors: {
      request: { use: vi.fn() },
      response: { use: vi.fn() },
    },
  };
  return { default: mockApi };
});

import api from '../services/api';
const mockApi = api as unknown as {
  post: ReturnType<typeof vi.fn>;
  get: ReturnType<typeof vi.fn>;
};

// ── Helpers ───────────────────────────────────────────────────────────────────
const mockUser = {
  firebaseUid: 'uid-001',
  firstName: 'Juan',
  lastName: 'dela Cruz',
  email: 'juan@test.com',
  phoneNumber: '',
  role: 'CUSTOMER' as const,
  isActive: true,
  token: 'mocked.jwt.token',
};

beforeEach(() => {
  localStorage.clear();
  vi.clearAllMocks();
});

afterEach(() => {
  localStorage.clear();
});

// ── Tests ─────────────────────────────────────────────────────────────────────

describe('authService', () => {
  describe('login()', () => {
    it('stores the JWT token and returns the user on success', async () => {
      mockApi.post.mockResolvedValueOnce({
        data: { data: mockUser },
      });

      const result = await authService.login({ email: 'juan@test.com', password: 'password123' });

      expect(result.email).toBe('juan@test.com');
      expect(localStorage.getItem('jwt_token')).toBe('mocked.jwt.token');
    });

    it('throws an error when the backend returns an error message', async () => {
      mockApi.post.mockRejectedValueOnce({
        response: { data: { error: 'Invalid email or password' } },
      });

      await expect(
        authService.login({ email: 'bad@test.com', password: 'wrong' })
      ).rejects.toThrow('Invalid email or password');
    });
  });

  describe('register()', () => {
    it('stores the JWT token and returns the user on success', async () => {
      mockApi.post.mockResolvedValueOnce({
        data: { data: mockUser },
      });

      const result = await authService.register({
        firstName: 'Juan',
        lastName: 'dela Cruz',
        email: 'juan@test.com',
        password: 'password123',
      });

      expect(result.email).toBe('juan@test.com');
      expect(localStorage.getItem('jwt_token')).toBe('mocked.jwt.token');
    });
  });

  describe('getCurrentUser()', () => {
    it('returns the current user from /auth/me', async () => {
      mockApi.get.mockResolvedValueOnce({ data: { data: mockUser } });

      const result = await authService.getCurrentUser();
      expect(result.firebaseUid).toBe('uid-001');
    });

    it('throws when the request fails', async () => {
      mockApi.get.mockRejectedValueOnce({
        response: { data: { error: 'Unauthorized' } },
      });

      await expect(authService.getCurrentUser()).rejects.toThrow('Unauthorized');
    });
  });

  describe('token helpers', () => {
    it('setToken / getToken / removeToken round-trip', () => {
      authService.setToken('test-token');
      expect(authService.getToken()).toBe('test-token');
      authService.removeToken();
      expect(authService.getToken()).toBeNull();
    });

    it('isAuthenticated() returns true when a token is stored', () => {
      authService.setToken('some-token');
      expect(authService.isAuthenticated()).toBe(true);
    });

    it('isAuthenticated() returns false when no token is stored', () => {
      expect(authService.isAuthenticated()).toBe(false);
    });
  });
});
